docs/configuration.md
System Configuration
==================

### Generating SSL Keys and Where They are Stored
Rabbitmq and MongoDB are configured to use SSL certificates to secure their communications. By default, during the setup of cyberCommons, these certificates are configured to be valid for 365 days. Once the certificates expire, they will need to be regenerated. If the ssl self-signed option was selected during installation, the certificates generated for the web server (NGINX) will also expire at this time and will need to be regenerated.

#### Generating SSL certificates
  Self-signed certificates are automatically generated on first run for RabbitMQ and MongoDB. Self-signed certificates for NGINX will be generated automatically during configuration if the self-signed option was selected during configuration.

  LetsEncrypt - refer to the [LetsEncrypt](installation.html#build-let-s-encrypt-docker-container) section of the installation instructions.

#### Renewing SSL Certificates
  1. Self-signed certificates can be updated by running the following command from the cyberCommons root directory:

  	$ run/genSSLKeys


  *All remote Celery workers will need the new SSL client certificates to resume communications. See the section below on where these certificates are stored*

  1. LetsEncrypt certificates can be renewed by running the following from the cyberCommons root directory:
  ~~~~
  $ config/ssl/nginx/runLetsEncrypt
  ~~~~

  *Follow LetsEncrypt's prompts*

#### SSL Certificate Locations
  1. Self-signed locations:
     * MongoDB
       - config/ssl/backend/client/mongodb.pem
       - config/ssl/backend/server/mongodb.pem
       - config/ssl/testca/cacert.pem
     * NGINX
       - config/ssl/nginx/keys/dhparam.pem
       - config/ssl/nginx/keys/selfsigned.crt
       - config/ssl/nginx/keys/selfsigned.key
     * RabbitMQ
       - config/ssl/backend/client/key.pem
       - config/ssl/backend/client/cert.pem
       - config/ssl/backend/server/key.pem
       - config/ssl/backend/server/cert.pem
       - config/ssl/testca/cacert.pem

  1. LetsEncrypt location:
     * NGINX
       - config/ssl/nginx/keys/dhparam.pem
       - config/ssl/nginx/letcencrypt/etc/live/*

### Configure Email Backend
* Uncomment and configure the Email Configuration section in config/api_config.py. *The following is an example using gmail.*
~~~
#*********** Email Configuration ********************
# Uncomment and configure to enable send email
EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
EMAIL_HOST = 'smtp.gmail.com'
EMAIL_PORT = 587
EMAIL_HOST_USER = 'username@gmail.com'
EMAIL_HOST_PASSWORD = 'password'
EMAIL_USE_TLS = True
~~~

* To enable Admin email notifications, add a line similar to the following to the config/api_config.py file. *Refer to Django's [documentation](https://docs.djangoproject.com/en/1.8/topics/email/) for more details on configuring email.*
~~~
ADMINS = [('Jane', 'ccadmin@your.org'), ('John', 'ccadmin2@your.org')]
~~~

### Add Admin User to MongoDB
1. Uncomment and configure the following section in the config/config.sh file. *Refer to MongoDB's [documentation](https://docs.mongodb.com/manual/reference/built-in-roles/) for more details on built-in-roles.*
~~~
# uncomment and populate the following two lines for the resetDBCreds application to add / update the mongo user admin account
mongo_admin_username=<adminuser>
mongo_admin_password=<use_strong_password>
mongo_admin_role=root
~~~

1. Run the following command from cyberCommons' root directory. *This step will restart the MongoDB server and create or update the admin user.*
~~~
$ run/resetDBCreds
~~~

### Open RabbitMQ and MongoDB Ports for Remote Workers
1. Edit run/cybercom_up with the following changes
  * Edit MongoDB config to include port 27017
  ~~~
  docker run -d -p 27017:27017 --name example_mongo \
  ~~~

  * Edit RabbitMQ config to include port 5671
  ~~~
  #Rabbitmq
  echo "************** Rabbitmq        ***********"
  docker run -d -p 5671:5671 --name example_rabbitmq \
  ~~~
1. Update firewall rules on host to allow access to these ports

### Turn Off Debug Mode for RESTful API
1. Set DEBUG = False in config/api_config.py
1. Add host(s) to ALLOWED_HOSTS list. See Django's documentation on the [ALLOWED_HOSTS](https://docs.djangoproject.com/en/1.8/ref/settings/#std:setting-ALLOWED_HOSTS) setting for more detail.


docs/configuration.rst
System Configuration
====================

Generating SSL Keys and Where They are Stored
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Rabbitmq and MongoDB are configured to use SSL certificates to secure
their communications. By default, during the setup of cyberCommons,
these certificates are configured to be valid for 365 days. Once the
certificates expire, they will need to be regenerated. If the ssl
self-signed option was selected during installation, the certificates
generated for the web server (NGINX) will also expire at this time and
will need to be regenerated.

Generating SSL certificates
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Self-signed certificates are automatically generated on first run for
RabbitMQ and MongoDB. Self-signed certificates for NGINX will be
generated automatically during configuration if the self-signed option
was selected during configuration.

LetsEncrypt - refer to the `LetsEncrypt <installation.html#build-let-s-encrypt-docker-container>`__
section of the installation instructions.

Renewing SSL Certificates
^^^^^^^^^^^^^^^^^^^^^^^^^

1. Self-signed certificates can be updated by running the following
   command from the cyberCommons root directory:

::

    $ run/genSSLKeys

*All remote Celery workers will need the new SSL client certificates to
resume communications. See the section below on where these certificates
are stored*

2. LetsEncrypt certificates can be renewed by running the following from
   the cyberCommons root directory:

::

    $ config/ssl/nginx/runLetsEncrypt

*Follow LetsEncrypt's prompts*

SSL Certificate Locations
^^^^^^^^^^^^^^^^^^^^^^^^^

1. Self-signed locations:

   -  MongoDB

      -  config/ssl/backend/client/mongodb.pem
      -  config/ssl/backend/server/mongodb.pem
      -  config/ssl/testca/cacert.pem
   -  NGINX

      -  config/ssl/nginx/keys/dhparam.pem
      -  config/ssl/nginx/keys/selfsigned.crt
      -  config/ssl/nginx/keys/selfsigned.key
   -  RabbitMQ

      -  config/ssl/backend/client/key.pem
      -  config/ssl/backend/client/cert.pem
      -  config/ssl/backend/server/key.pem
      -  config/ssl/backend/server/cert.pem
      -  config/ssl/testca/cacert.pem

2. LetsEncrypt location:

   -  NGINX

      -  config/ssl/nginx/keys/dhparam.pem
      -  config/ssl/nginx/letcencrypt/etc/live/\*

Configure Email Backend
~~~~~~~~~~~~~~~~~~~~~~~

-  Uncomment and configure the Email Configuration section in
   config/api\_config.py. *The following is an example using gmail.*

.. code-block:: python

    #*********** Email Configuration ********************
    # Uncomment and configure to enable send email
    EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
    EMAIL_HOST = 'smtp.gmail.com'
    EMAIL_PORT = 587
    EMAIL_HOST_USER = 'username@gmail.com'
    EMAIL_HOST_PASSWORD = 'password'
    EMAIL_USE_TLS = True

-  To enable Admin email notifications, add a line similar to the
   following to the config/api\_config.py file. Refer to Django's
   `documentation <https://docs.djangoproject.com/en/1.8/topics/email/>`__
   for more details on configuring email.

.. code-block:: python

   ADMINS = [('Jane', 'ccadmin@your.org'), ('John', 'ccadmin2@your.org')]

Add Admin User to MongoDB
~~~~~~~~~~~~~~~~~~~~~~~~~

1. Uncomment and configure the following section in the config/config.sh
   file. Refer to MongoDB's
   `documentation <https://docs.mongodb.com/manual/reference/built-in-roles/>`__
   for more details on built-in-roles.

.. code-block:: bash

    # uncomment and populate the following two lines for the resetDBCreds application to add / update the mongo user admin account
    mongo_admin_username=<adminuser>
    mongo_admin_password=<use_strong_password>
    mongo_admin_role=root

2. Run the following command from cyberCommons' root directory. *This
   step will restart the MongoDB server and create or update the admin
   user.*

::

    $ run/resetDBCreds

Open RabbitMQ and MongoDB Ports for Remote Workers
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. Edit run/cybercom_up with the following changes

  -  Edit MongoDB config to include port 27017

  .. code-block:: bash

    docker run -d -p 27017:27017 --name example_mongo \

  -  Edit RabbitMQ config to include port 5671

  .. code-block:: bash

    #Rabbitmq
    echo "************** Rabbitmq        ***********"
    docker run -d -p 5671:5671 --name example_rabbitmq \

2. Update firewall rules on host to allow access to these ports

Turn Off Debug Mode for RESTful API
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. Set DEBUG = False in config/api_config.py
2. Add host(s) to ALLOWED_HOSTS list. See Django's documentation on the
   `ALLOWED_HOSTS <https://docs.djangoproject.com/en/1.8/ref/settings/#std:setting-ALLOWED_HOSTS>`__
   setting for more detail.


api/config.py
__author__ = 'cyberCommons Framework'

import os
import ssl
BASE_DIR = os.path.dirname(os.path.dirname(__file__))

# ****** Application Settings *******************************************************

APPLICATION_TITLE = os.environ.get('APPLICATION_TITLE', 'Cybercommons')

# ******* Backend Communication Celery Brooker and MongoDB **************************

MEMCACHE_HOST = "cybercom_memcache"
MEMCACHE_PORT = 11211

# Celery workers config are setup with the folling DB and Collections
# Change would reqire updating celeryconfig.py within celery image folder
MONGO_DB = os.environ.get('MONGO_DB', "cybercom")
MONGO_LOG_COLLECTION = os.environ.get('MONGO_LOG_COLLECTION', "task_log")
MONGO_TOMBSTONE_COLLECTION = os.environ.get('MONGO_TOMBSTONE_COLLECTION', "tombstone")

SSL_PATH = os.environ.get('SSL_PATH')
RABBITMQ_DEFAULT_USER = os.environ.get('RABBITMQ_DEFAULT_USER')
RABBITMQ_DEFAULT_PASS = os.environ.get('RABBITMQ_DEFAULT_PASS')
RABBITMQ_DEFAULT_VHOST = os.environ.get('RABBITMQ_DEFAULT_VHOST')
BROKER_URL = "amqp://{0}:{1}@cybercom_rabbitmq:5671/{2}"
BROKER_URL = BROKER_URL.format(RABBITMQ_DEFAULT_USER, RABBITMQ_DEFAULT_PASS, RABBITMQ_DEFAULT_VHOST)
BROKER_USE_SSL = {
    'keyfile': '{0}/client/key.pem'.format(SSL_PATH),
    'certfile': '{0}/client/cert.pem'.format(SSL_PATH),
    'ca_certs': '{0}/testca/cacert.pem'.format(SSL_PATH),
    'cert_reqs': ssl.CERT_REQUIRED
}
MONGO_USERNAME = os.environ.get('MONGO_USERNAME')
MONGO_PASSWORD = os.environ.get('MONGO_PASSWORD')
MONGO_HOST = os.environ.get('MONGO_HOST', 'cybercom_mongo')
MONGO_PORT = os.environ.get('MONGO_PORT', '27017')
MONGO_URI = f"mongodb://{MONGO_USERNAME}:{MONGO_PASSWORD}@{MONGO_HOST}:{MONGO_PORT}/?ssl=true&ssl_ca_certs={SSL_PATH}/testca/cacert.pem&ssl_certfile={SSL_PATH}/client/mongodb.pem"
CELERY_RESULT_BACKEND = MONGO_URI
CELERY_MONGODB_BACKEND_SETTINGS = {
    "database": MONGO_DB,
    "taskmeta_collection": MONGO_TOMBSTONE_COLLECTION
}
# ******* Catalog ******************************************************
CATALOG_EXCLUDE = ['admin', 'config', 'local', 'default_collection', MONGO_DB]
CATALOG_INCLUDE = ['catalog']
CATALOG_URI = MONGO_URI
CATALOG_ANONYMOUS = True
# *********** Data Store ************************************************
DATA_STORE_EXCLUDE = ['admin', 'config', 'local', 'default_collection', MONGO_DB, 'catalog']
DATA_STORE_MONGO_URI = MONGO_URI
DATA_STORE_ANONYMOUS = True

# If you want to enforce perms for SAFE_METHOD
# Set SAFE_METHOD_PERM_REQUIRED Env Variable
# Example: "mydb_mycollection,mydb1_mycollection1"
SAFE_METHOD_PERM_REQUIRED= os.environ.get('SAFE_METHOD_PERM_REQUIRED','').split(',')

# *********** Email Configuration ********************
if os.getenv('EMAIL_HOST'):
    EMAIL_BACKEND = os.getenv('EMAIL_BACKEND')
    EMAIL_HOST = os.getenv('EMAIL_HOST')
    EMAIL_PORT = os.getenv('EMAIL_PORT')
    EMAIL_HOST_USER = os.getenv('EMAIL_HOST_USER')
    EMAIL_HOST_PASSWORD = os.getenv('EMAIL_HOST_PASSWORD')
    EMAIL_USE_TLS = os.getenv('EMAIL_USE_TLS', 'false').lower() == 'true'

